### ISSUE TITLE
Qr Scanner wrapping

### ISSUE THREAD

**Issue by ZeD4805** [reflex-dev/reflex#1605]
I'd like to develop a wrap of an npm package to scan qr codes but my knowledge of React/React-native (???) is quite limited. I've tried somethings but I haven't really gone anywhere with it.

I think this can be quite useful for those developing anything for mobile phones on reflex.

The package is [qr-scanner](https://github.com/nimiq/qr-scanner).

Thank you in advance for any tips/help.

(In some discord threads it was suggested to open an issue if doing the wrap myself was too much for my React knowledge).

I've included the library in the code on the rxconfig.py:
```
    frontend_packages=[
        "qr-scanner",
    ],
```

And also created the component class
```
from typing import Optional
import reflex as rx

class QrScanner(rx.Component):
    """Wrapper for the QrScanner react nmp package"""
    library = "qr-scanner"
    tag = "QrScanner"
    is_default=True
    result: rx.Var[str]

    @classmethod
    def get_controlled_triggers(cls) -> dict[str, rx.Var]:
        return {"on_change": rx.EVENT_ARG}
    
qrscanner = QrScanner.create
```

And instanced it on the index component:
```
import reflex as rx

from reflex_qr_scanner_npm.qr_scanner import qrscanner, QrScanner

class State(rx.State):
    """The app state."""

    result: str

def index() -> rx.Component:
    return rx.fragment(
        rx.vstack(
            rx.heading("Qr scanner", font_size="2em"),
            rx.video(
                
            ),
            qrscanner(
                result=State.result,
                on_change=lambda result: State.set_result(result),
            ),
        ),
    )


# Add state and page to the app.
app = rx.App()
app.add_page(index)
app.compile()

```

**Comment by masenf** [comment]
did you add the library to `frontend_packages` in `rxconfig.py`?

similar to #1352

**Comment by ZeD4805** [comment]
I did, it's the first code block in the issue :)

I was actually not aware of the issue you referred, I was about to close my own but I've tested the example you tried and I get a react error when rendering the page:
```
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.

Check the render method of `ForwardRef(LoadableComponent)`.
```

I have not changed any of the code you've provided. This is my rxconfig.py:

```
import reflex as rx

class ReflexqrscannernpmConfig(rx.Config):
    pass

config = ReflexqrscannernpmConfig(
    app_name="reflex_qr_scanner_npm",
    api_url="http://localhost:8000",
    frontend_packages=[
        "@yudiel/react-qr-scanner",
    ],
)
```

**Comment by masenf** [comment]
Ah, okay. It looks like when I made that example, the `NoSSRComponent` code was different. The example actually depends on [this PR](https://github.com/reflex-dev/reflex/pull/1548), which will be in the _next_ release.

For wrapping things that use client-side features, it is necessary to use the `NoSSRComponent` to avoid server side rendering that would otherwise occur by default.

Can you try the example code with the tip of `main`. I was able to get it working without modifications tonight.


https://github.com/reflex-dev/reflex/assets/1524005/e4a8b2c7-b957-412c-b28b-c06fc24c0e38

**Comment by masenf** [comment]
I also took a closer look at the library you were intending to wrap and unfortunately the way it's designed makes it difficult to integrate with reflex because it's not already wrapped as a react component and the configuration is largely imperative. Wrapping it wouldn't be impossible, but it would require more work and more custom JS code to actually turn it into a react component that could be used in the page.

If you're able to get the `@yudiel/react-qr-scanner` working, i think that will be less code overall. I tried wrapping three different qr scanners a couple weeks ago and the `yudiel` one definitely worked the best in my testing.

**Comment by ZeD4805** [comment]
> Can you try the example code with the tip of `main`. I was able to get it working without modifications tonight.

Can confirm it works from `main`! Thank you very much!

**Comment by ZeD4805** [comment]
@masenf 
I am having another issue, if I wrap the component around a rx.cond component, and I try to close the container, I get the following error message:

```
TypeError: p.current is not a function

Call Stack
eval
node_modules/@yudiel/react-qr-scanner/dist/esm/index.js (1:394540)
n
node_modules/@yudiel/react-qr-scanner/dist/esm/index.js (1:53299)
```

```python
import reflex as rx

from .yudiel_qr_scanner import qr_scanner as yudiel_qr_scanner


class State(rx.State):
    scanning: bool = False
    last_scan: str = ""

    def on_scan(self, result: str):
        print(f"on_scan: {result}")
        if result:
            self.last_scan = result
            #self.scanning = False


def yudiel_react_qr_scanner_page() -> rx.Component:
    return rx.fragment(
        rx.vstack(
            rx.heading("@yudiel/react-qr-scanner", font_size="2em"),
            rx.box(
                rx.cond(
                    State.last_scan,
                    rx.text(State.last_scan),
                    rx.text("Scan a valid QR code"),
                ),
            ),
            rx.cond(
                State.scanning,
                rx.vstack(
                    yudiel_qr_scanner(
                        on_decode=State.on_scan,
                        container_style={
                            "width": "640px",
                            "height": "480px",
                            "padding-top": "0",
                        },
                        scan_delay=500,
                    ),
                    rx.button(
                        "Stop scanning",
                        on_click=State.set_scanning(False),
                    ),
                ),
                rx.button(
                    "Scan",
                    on_click=State.set_scanning(True),
                ),
            ),
            spacing="1.5em",
            font_size="2em",
            padding_top="10%",
        ),
    )


app = rx.App()
app.add_page(yudiel_react_qr_scanner_page, route="/")
app.compile()
```

Am I missing anything to close the container?

**Comment by masenf** [comment]
To pass a value to an event handler, you can use a lambda like

`on_click=lambda: State.set_scanning(True)`

Not sure if that's the only problem, but that's what jumps out to me immediately. I can actually try out your sample code in a few hours.

**Comment by ZeD4805** [comment]
That's an absolutely valid point, however, even after correcting that it still reports the same issue.

**Comment by ZeD4805** [comment]
@masenf I'm not sure you were tagged for my last message.


> That's an absolutely valid point, however, even after correcting that it still reports the same issue.

Are there other things you that you can see are wrong?

**Comment by Lendemor** [comment]
I did some testing on my side, for me the error occurs regardless of using rx.cond or not, so I don't think it is related.

**Comment by Lendemor** [comment]
Okay I may have figured out the reason for the Unhandled Runtime Error.

By giving an on_error event handler to the component, we avoid the traceback.

```python
import reflex as rx

from .yudiel_qr_scanner import qr_scanner as yudiel_qr_scanner


class State(rx.State):
    scanning: bool = False
    last_scan: str = ""

    def on_scan(self, result: str):
        print(f"on_scan: {result}")
        if result:
            self.last_scan = result

    def on_error(self, error: str):
        print(f"on_error: {error}")


def yudiel_react_qr_scanner_page() -> rx.Component:
    return rx.fragment(
        rx.vstack(
            rx.heading("@yudiel/react-qr-scanner", font_size="2em"),
            rx.box(
                rx.cond(
                    State.last_scan,
                    rx.text(State.last_scan),
                    rx.text("Scan a valid QR code"),
                ),
            ),
            yudiel_qr_scanner(
                on_decode=State.on_scan,
                on_error=State.on_error,
                container_style={
                    "width": "640px",
                    "height": "480px",
                    "paddingTop": "0",
                },
                scan_delay=500,
            ),
            spacing="1.5em",
            font_size="2em",
            padding_top="10%",
        ),
    )


app = rx.App()
app.add_page(yudiel_react_qr_scanner_page, route="/")
app.compile()
```

I didn't tested it on a phone to see if it's actually working with a camera however.

**Comment by ZeD4805** [comment]
@Lendemor It definitely works now, but is there a reason it just throws an empty error at every scan?

**Comment by GHC19** [comment]
Do you have any code for doing this wrapping react for scanning ?

**Comment by masenf** [comment]
@GHC19 we did have a working example, but it didn't get updated for the latest reflex versions or the new API of `react-qr-scanner`.

I made a PR to update it to the latest conventions (although still using chakra components): https://github.com/reflex-dev/reflex-chakra/pull/28

That depends on currently unreleased features though, so it will take some time to get it in and working again, but i'll keep an eye on it.

Ultimately someone should publish a third party component that wraps the 2.x version of the API and then it would be very easy to include this functionality in a multitude of reflex apps.
